src/merkle/path_set.rs

```diff
@@ -111,18 +111,42 @@ impl MerklePathSet {
             });
         }
 
-        let index_value = index.to_scalar_index();
-        let index = index_value / 2;
-        let parity = index_value & 1;
+        let parity = index.value() & 1;
+        let path_key = index.value() - parity;
         let mut path = self
             .paths
-            .get(&index)
+            .get(&path_key)
             .cloned()
-            .ok_or(MerkleError::NodeNotInSet(index))?;
+            .ok_or(MerkleError::NodeNotInSet(index.value()))?;
         path.remove(parity as usize);
         Ok(path)
     }
 
+    /// Returns all paths in this path set together with their indexes.
+    pub fn to_paths(&self) -> Vec<(u64, ValuePath)> {
```

Comment:
optional nit: here we can do a more generic approach with:

```rust
pub fn to_paths(&self) -> impl Iterator<Item = (u64, ValuePath)> + '_
```

The advantage of this is to leave more options for the user on how to take the input. Sometimes, as on the `MerkleStore::add_merkle_paths`, the user won't need this newly collected vec, making the allocation unnecessary.

Another nice to have would be to avoid tuples as return. They are quite hard to keep documented and aren't immediately clear on the purpose. We could have something like

```rust
/// A container for a leaf index and its [MerklePath].
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct IndexPath {
    /// The index of the opening
    pub index: u64,
    /// The path from `leaf` to `root` (exclusive).
    pub path: MerklePath,
}
```

(or even use `NodeIndex` to make it brutally clear we always use the same depth; that would be redundant tho because the depth is already embedded onto the [MerklePath] itself via its length)

Reply:
Yep - makes sense, though I think this would require us creating a custom iterator struct. I'd rather keep it simple for now as I think we might get rid of the `MerklePathSet` entirely in the future.